### PR TITLE
Add check timer sources debug command

### PR DESCRIPTION
**Pull request by Mothblocks**
## About The Pull Request
Adds a "Check Timer Sources" command that allows those with debug access to check the sources of commands.

Turns `/proc/addtimer` into `/proc/_addtimer` and creates an `addtimer` macro in its place. This is because the macro will pass in the file and line of what is creating the timer. I wanted to use `gib_stack_trace()` and friends, but those regularly turned up blank exception descriptions, and were useless. If there's anything better, let me know, but for now this is better than nothing.

The command will only show those still in the buckets. This is to avoid both leaving too much of a footprint on the `addtimer` proc and due to memory concerns of logging all timers being created.

image

Based on a request by @optimumtact, let me know if this is not what you are looking for.

## Why It's Good For The Game

Bagil is having issues of unplayable lag. Maintainers have reason to believe it is due to something creating an absurd amount of timers. This should help isolate the issue.

## Changelog
:cl:
admin: Added the "Check Timer Sources" debug command to help isolate problematic cases of addtimer.
/:cl:

### CODE REVIEW

**Comment by optimumtact** [comment]
good good, contact me on discord to sort out payment, paypal is easiest

**Comment by AnturK** [comment]
Good PR

File: code/modules/admin/verbs/debug.dm [hunk 1]
```diff
@@ -889,3 +889,59 @@ But you can call procs that are of type /mob/living/carbon/human/proc/ for that
...
+/// A debug verb to check the sources of currently running timers
+/client/proc/check_timer_sources()
+	set category = "Debug"
+	set name = "Check Timer Sources"
+	set desc = "Checks the sources of the running timers"
+	if (!check_rights(R_DEBUG))
+		return
+
+	var/bucket_list_output = generate_timer_source_output(SStimer.bucket_list)
+	var/second_queue = generate_timer_source_output(SStimer.second_queue)
```
Comment (MrStonedOne) [inline code/modules/admin/verbs/debug.dm:902]: bucket_list is a fixed sized list of linked lists.

Meaning each timer in the bucket list is only the HEAD of a linked list. `generate_timer_source_output` here is only looking at the head timer and not any other item in the linked list.

next/prev vars will be null on secondqueue, so you could safely add a linked list transversal to generate_timer_source_output to fix this.

see reset_buckets for code designed to loop through all timers https://github.com/tgstation/tgstation/blob/master/code/controllers/subsystem/timer.dm#L236

File: code/modules/admin/verbs/debug.dm [hunk 2]
```diff
@@ -889,3 +889,59 @@ But you can call procs that are of type /mob/living/carbon/human/proc/ for that
...
+	// Collate all events and figure out what sources are creating the most
+	for (var/_event in events)
+		if (!_event)
+			continue
+		var/datum/timedevent/event = _event
+		if (event.source)
+			if (per_source[event.source] == null)
+				per_source[event.source] = 1
+			else
+				per_source[event.source] += 1
```
Comment (MrStonedOne) [inline code/modules/admin/verbs/debug.dm:924]:
```suggestion
	for (var/_event in events)
		if (!_event)
			continue
		var/datum/timedevent/event = _event
		do
			if (event.source)
				per_source[event.source]++
			event = event.next
		while(event && event != _event) 
```

**Comment by Jared-Fogle** [comment]
Thanks for the suggestion. That made it recognize a lot more timers.

image

**Comment by TemporalOroboros** [comment]
Wow materials are creating a lot of timers.

https://github.com/tgstation/tgstation/blob/0d66933e0895872e1928c5a8814ad63ec76ba4e4/code/datums/materials/_material.dm#L47-L69

???

**Comment by Jared-Fogle** [comment]
@TemporalOroboros I looked into it a little, that's called through `/atom/Initialize` and creates a new timer for *every* atom with a material. I'm really not sure.

**Comment by TemporalOroboros** [comment]
It isn't just creating a timer for every atom containing the relevant materials... It's creating a timer for every one of the relevant materials when the material is added to an atom. And all of the atoms are added individually...

The relevant materials are:
- Glass
- Silver
- Gold
- Diamond
- Uranium
- Plasma
- Bluespace Crystal
- Bananium (Who else has been pronouncing this banananium? Just me?)
- Titanium
- Runite
- Plastic
- Wood
- Adamantine
- Mythril
- Hot Ice
- Metallic Hydrogen
- Sand
- Sandstone
- Snow
- Runed Metal
- Bronze
- Paper
- Cardboard
- Bone
- Bamboo
- Pizza
- Meat
- And any subtypes of the above...

**Comment by Jared-Fogle** [comment]
@Qustinnus It seems you were the one that added that, mind giving some input as to why its written the way it is?

**Comment by AnturK** [comment]
Conflicts

**Comment by optimumtact** [comment]
I love you bro

**Pull request merged**